### ISSUE TITLE
RFC: enhancement: native RobustIRC support

### ISSUE THREAD

**Issue by stapelberg** [weechat/weechat#385]
I’m one of the developers of http://robustirc.net/. See http://robustirc.net/docs.html for more details, but in a nutshell, it’s an IRC network written on top of https://raftconsensus.github.io/ which doesn’t suffer from netsplits. To not be affected by network issues between the client and the IRC servers, clients should use the RobustSession protocol which wraps IRC in a connection-independent HTTPS/JSON-based transport, see http://robustirc.net/docs/robustsession.html for details.

Currently, we offer a program called the RobustIRC bridge, which you can configure as a proxy server in WeeChat. The bridge then does the conversion (bridging) from IRC to IRC in RobustSession.

I intend to submit a pull request for WeeChat to talk RobustSession natively, i.e. without the bridge. Would such a pull request be accepted?

Thank you.

**Comment by flashcode** [comment]
Hi Michael,
First thanks for i3, I love your window manager ;)
I didn't read everything about RobustSession yet, but if I understand correctly it would be an encapsulation of IRC message in another protocol, so it should impact almost only the way to send/receive messages in irc plugin (network stuff), am I right?
Feel free to send a pull request, and we'll discuss there about the changes.

**Comment by Shawn-Smith** [comment]
I skimmed through the docs a bit. It's quite a big change and I think it'd be better off as a new plugin rather than added to the irc plugin.

**Comment by stapelberg** [comment]
> First thanks for i3, I love your window manager ;)

Glad to hear you like it :).

> I didn't read everything about RobustSession yet, but if I understand correctly it would be an encapsulation of IRC message in another protocol, so it should impact almost only the way to send/receive messages in irc plugin (network stuff), am I right?

You’re correct — it’s only a transport layer, and the client still speaks IRC over it just as it does today.

> I skimmed through the docs a bit. It's quite a big change and I think it'd be better off as a new plugin rather than added to the irc plugin.

We briefly evaluated this before (by just reading the WeeChat plugin docs) and didn’t see an easy way to say that our protocol should be based on the IRC plugin, just with a different transport. Can you outline how we’d go about implementing this in a separate plugin please?

**Comment by Mikaela** [comment]
Is RobustIRC official standard?

The next version of IRC is [IRCv3](http://ircv3.org) and it's development is happening at \#ircv3 on chat.freenode.net. You might want to contact them.

**Comment by kaniini** [comment]
As an aside, what is the point of encapsulating IRC frames inside JSON inside HTTP?  How does this "improve" robustness?

**Comment by stapelberg** [comment]
@Mikaela Depends on how you define “standard”. We can certainly write an RFC for the RobustSession protocol.

@kaniini Please see http://robustirc.net/docs.html, especially the presentation. The RobustSession protocol is merely necessary, not sufficient, to achieve higher robustness.

**Comment by SaberUK** [comment]
From what I can see RobustIRC seems like an attempt to fix IRC without any understanding of what is actually wrong with IRC.

**Comment by stapelberg** [comment]
@SaberUK Without any further clarification, your comment just comes off as arrogant to me. Can you point out what you think should have been adressed instead, or how RobustIRC should be different?

**Comment by SaberUK** [comment]
@stapelberg You are trying to solve netsplits by adding eleventy layers of complexity on top of the existing protocol instead of doing the logical thing which is mesh linking.

**Comment by stapelberg** [comment]
@SaberUK Is there a description/specification of the S2S mesh linking you’re referring?

**Comment by kaniini** [comment]
@stapelberg Why is RobustSession necessary at all for improving robustness?  Why the specific composition of HTTP and JSON?  Also, why the necessity for one complete HTTP GET request per IRC frame?  What if there are multiple events in the same timeslice?  How does RobustIRC server handle actively hostile nodes w.r.t. raft consensus?

**Comment by kaniini** [comment]
I believe @SaberUK is advocating a theoretical "fabric" topology where nodes maintain a routing table to resolve topological cycles.  Something akin to IS-IS or BGP, except at the IRC protocol layer.

**Comment by stapelberg** [comment]
> Why is RobustSession necessary at all for improving robustness?

So that individual TCP connection failures are irrelevant.

> Why the specific composition of HTTP and JSON?

Because HTTP and JSON are very widely available building blocks.

> Also, why the necessity for one complete HTTP GET request per IRC frame?

That’s not how it works. The HTTP GET is long-running and a single IRC message is sent as an HTTP chunk. In practice, a single HTTP GET will transport hundreds of IRC messages.

> What if there are multiple events in the same timeslice?

What do you mean by timeslice?

> How does RobustIRC server handle actively hostile nodes w.r.t. raft consensus?

It doesn’t. You need to protect your network from actively malicious nodes, raft doesn’t protect you against that.

Let me know if any of my explanations were unclear and I’ll try to be more verbose (ideally you’d ask a more specific question).

**Comment by kaniini** [comment]
> So that individual TCP connection failures are irrelevant.
> Because HTTP and JSON are very widely available building blocks.

I think you misunderstand what I am saying.  You're transporting naked rfc1459/2812 frames inside JSON.  Why bother with all of this abstraction?  What is the point?  Why not at least parse the frames into fragments so you can validate they are not malformed?  Further, I do not think you really need HTTP overhead for this, why not just use sockets?

> What do you mean by timeslice?

A moment in time, such as between 1426109846660183024.1 and 1426109846660183024.2, since your timestamping for events is based on microseconds. (although this seems odd, because timestamping itself is not needed for message ordering, if your primitives are sufficiently atomic.)

> You need to protect your network from actively malicious nodes, raft doesn’t protect you against that.

Oh.  Okay then.  I'm sorry, but this alone means it's less robust than actual modern IRC linking primitives... IRC linking protocols have both consensus resolution and correctness proofs through UUIDs and timestamps (although this could stand to be further improved).  In a RobustIRC network, there is apparently no way to handle a badly behaved node operator from overriding access primitives on the various objects without permission?  I thought raft was supposed to enforce consensus in this regard?

About this, by the way, raft is still affected by network partitions (netsplits).  Raft is actually a *downgrade* because it only *safely* handles primary-secondary failover without data loss.  Any other condition (including hostile nodes as you just said) can result in data loss.  In fact, see Aphyr's [post on etcd and Consul as an example of known issues with raft](https://aphyr.com/posts/316-call-me-maybe-etcd-and-consul).  Your description of Raft as being a cure for netsplits makes me nervous, because it's less of a cure than current linking protocols which [guarantee consistency once a network partition is healed in all cases](http://www.stack.nl/~jilles/irc/mode-handling.txt).

**Comment by stapelberg** [comment]
Thanks for the thoughtful post.

> I think you misunderstand what I am saying. You're transporting naked rfc1459/2812 frames inside JSON. Why bother with all of this abstraction? What is the point?

Well, we’re not _only_ transporting RFC1459/2812 messages, but also other properties (such as the ID of the message, or, in the case of a robustPing message, the current list of servers). Once you’re transporting more than one piece of data, you somehow need to serialize it, and we chose JSON for that due to it being very widely available.

> Further, I do not think you really need HTTP overhead for this, why not just use sockets?

Of course just using sockets would be possible, but I think HTTP is rather well-known and makes it easy to reason about the RobustSession protocol in terms of requests/responses, parameters, headers (such as the per-session shared secret) and error codes, all of which we would otherwise need to re-invent.

> A moment in time, such as between 1426109846660183024.1 and 1426109846660183024.2, since your timestamping for events is based on microseconds. (although this seems odd, because timestamping itself is not needed for message ordering, if your primitives are sufficiently atomic.)

Actually, the part after the dot is the reply id and not part of the timestamp (a message resulting in two replies from the server will result in these replies having the .1 and .2 id).

While we _do_ currently use timestamps, the only requirement for the IDs is that they are strictly monotonically increasing. This is necessary for clients resuming (i.e. sending a GetMessages request with lastseen=) after being disconnected from the network.

>> You need to protect your network from actively malicious nodes, raft doesn’t protect you against that.
>
> Oh. Okay then. I'm sorry, but this alone means it's less robust than actual modern IRC linking primitives... IRC linking protocols have both consensus resolution and correctness proofs through UUIDs and timestamps (although this could stand to be further improved).

“less robust” depends on your definition of robustness, which I realize is a subjective term that can be applied in an overly broad sense :). Let’s make it more concrete for this discussion: the “robust” in “RobustIRC” means there are no netsplits when there is partial server/network unavailability.

> In a RobustIRC network, there is apparently no way to handle a badly behaved node operator from overriding access primitives on the various objects without permission?

How would a node operator do that specifically? The only feasible way I can think of involves manipulating the memory of the running process (e.g. with GDB), which you can also just do with any other IRC server/program.

> About this, by the way, raft is still affected by network partitions (netsplits).

Can you elaborate on this please? We have been running a RobustIRC based network for over three months now and I can assure that we haven’t had a traditional netsplit, i.e. server updates/restarts/network unavailability actually do not cause netsplits.

> Raft is actually a downgrade because it only safely handles primary-secondary failover without data loss.

There is no concept of “primary” and “secondary” in Raft. I believe the terminology you’re looking for is “leader” and “follower”. There can only be one leader at any point in time, but multiple followers.

> Any other condition (including hostile nodes as you just said) can result in data loss.

I think hostile nodes are the _only_ condition that can result in data loss. To be clear: data loss in this case means that you need to force a traditional netsplit by turning down all nodes and deleting their state.

> Your description of Raft as being a cure for netsplits makes me nervous, because it's less of a cure than current linking protocols which guarantee consistency once a network partition is healed in all cases.

The big difference between RobustIRC (or a raft-based network in general) and current networks is that netsplits are handled transparently, whereas, while current linking protocols achieve consistency after a netsplit, that will be very visible and impactful to end users.

**Comment by kaniini** [comment]
Watching your presentation again explains it better but gives me other concerns.

Allowing the network to break on an orphaned node seems unacceptable, while netsplits are harmful, for the most part, IRCv3.2 batching allows for the visual effect to be suppressed.  In this state, an orphaned node is still functional as an IRC server which may be useful in some edge cases.

Furthermore, a proposal for session resumption has been submitted for inclusion in IRCv3.3, which seems to handle the same use-case as `RobustSession`.  Also the next generation of linking protocols being proposed generally are log-structured, similar to Raft, so the lost messages can be potentially replayed upon the partition being healed.

Furthermore, you mention in the presentation that not all log events can be compacted, are you able to compact events which are obsoleted (such as old iterations of `TOPIC`)?

I also wonder why it's called RobustIRC since does not actually speak IRC protocol (without translation from `RobustSession`).

**Comment by stapelberg** [comment]
> Allowing the network to break on an orphaned node seems unacceptable, while netsplits are harmful, for the most part, IRCv3.2 batching allows for the visual effect to be suppressed. In this state, an orphaned node is still functional as an IRC server which may be useful in some edge cases.

I’m not entirely sure what your point is? Can you clarify?

> Furthermore, a proposal for session resumption has been submitted for inclusion in IRCv3.3, which seems to handle the same use-case as RobustSession. Also the next generation of linking protocols being proposed generally are log-structured, similar to Raft, so the lost messages can be potentially replayed upon the partition being healed.

That’s good to know. It sounds like when IRCv3.3 is implemented and widely supported in clients/servers, RobustIRC might be obsolete, which is fine :).

> Furthermore, you mention in the presentation that not all log events can be compacted, are you able to compact events which are obsoleted (such as old iterations of TOPIC)?

All obsolete events can be compacted, yes. 

> I also wonder why it's called RobustIRC since does not actually speak IRC protocol (without translation from RobustSession).

RobustSession is a transport protocol, not defining any semantics. The protocol which is used between clients and servers is in fact IRC, just transported over RobustSession.

**Comment by Shawn-Smith** [comment]
>We briefly evaluated this before (by just reading the WeeChat plugin docs) and didn’t see an easy way to say that our protocol should be based on the IRC plugin, just with a different transport. Can you outline how we’d go about implementing this in a separate plugin please?

@stapelberg In all honesty I'm not sure, I haven't written a plugin for WeeChat before. But since this is a new transport, that itself is quite a large change. I think it would be best to copy the irc plugin to a new name (robustirc?) and then start modifying that to fit your new transport needs.

The #weechat channel on freenode would likely be able to help with this more.


>That’s good to know. It sounds like when IRCv3.3 is implemented and widely supported in clients/servers, RobustIRC might be obsolete, which is fine :).

This is another reason for me wanting it in it's own plugin, if this is added to the IRC plugin and it becomes obsolete then WeeChat ends up with a bunch of dead code in the IRC plugin that virtually no one uses. However as it's own plugin then people who aren't using it can simply unload it and not have to worry about it.

**Comment by stapelberg** [comment]
I have gotten a proof of concept implementation to work. It currently replaces the code of  `irc_server_send()` with a couple of calls to curl and calls `irc_server_msgq_add_buffer()` from curl callbacks.

Given that these are both functions that are internal to the IRC plugin, I don’t think we can easily put my code into a separate plugin without some refactoring.

> This is another reason for me wanting it in it's own plugin, if this is added to the IRC plugin and it becomes obsolete then WeeChat ends up with a bunch of dead code in the IRC plugin that virtually no one uses. 

What exactly is your concern? If the code is not executed, it cannot expose bugs. If it turns out at some point that RobustIRC is dead and there are no public networks anymore, we can deprecated and then remove the RobustIRC-related code.

I think a better alternative would be to cleanly separate the RobustIRC-related code into its own file and call the functions exposed by that file in the existing IRC plugin.

@flashcode, what do you think?

**Comment by Shawn-Smith** [comment]
>What exactly is your concern? If the code is not executed, it cannot expose bugs. If it turns out at some point that RobustIRC is dead and there are no public networks anymore, we can deprecated and then remove the RobustIRC-related code.

Once it's there it can't easily be removed without negative feedback from people who may still use it, if other projects like RobustIRC turn up and support is added for them as well then the irc plugin is going to become bloated.

If RobustIRC is added as its own plugin and is depreciated sometime in the future the plugin can be kept for the select users that may still want to use it. Those that don't want to use it can simply opt not to load the plugin.

Dead code doesn't directly execute bugs, but it can contribute to bugs not being found. OpenSSL's Heartbleed is sort of an example of this: http://servicevirtualization.com/profiles/blogs/what-heartbleed-teaches-us-about-code-risk

>Beck: The existing code base is full of things that should have been deleted years ago. Dead code abounds and the maze of most of it makes things very hard to follow.

--
>I think a better alternative would be to cleanly separate the RobustIRC-related code into its own file and call the functions exposed by that file in the existing IRC plugin.

RobustIRC isn't IRC. That's not bad, but there's no reason to integrate it with the current IRC plugin when WeeChat was designed to be modular and allow for different plugins to do different transport protocols. For example here's a Tox plugin for WeeChat: https://github.com/haavardp/tox-weechat

**Comment by stapelberg** [comment]
> That's not bad, but there's no reason to integrate it with the current IRC plugin when WeeChat was designed to be modular and allow for different plugins to do different transport protocols.

I’m not convinced this statement is applicable to the case at hand. The example you name (tox) is an entirely different _protocol_, not just a new _transport_ for an existing protocol.

Earlier you suggested to just copy the IRC plugin, but I don’t think that’s a good alternative, because then we end up with a huge amount of code duplication.

The best way to make WeeChat use RobustSession would be to have a robustsession plugin that interacts with the irc plugin. As I already explained in my previous comment, I don’t see how that is possible with WeeChat’s current architecture. Can you please outline how (on a technical level) this is supposed to work?

**Comment by Mikaela** [comment]
I also think this should be separate plugin and same might apply to json in the future when IRCv3.2 brings it.

**Comment by stfnm** [comment]
@stapelberg With [weechat_hook_modifier()](https://weechat.org/files/doc/devel/weechat_plugin_api.en.html#_hook_modifier) it's possible to modify incoming messages (modifier "irc_in_xxx") as well as outgoing messages (modifier "irc_out_xxx"), would that be enough to do the job or is that too high level already?

**Comment by talisein** [comment]
@stapelberg Can RobustIRC be abstracted out as a proxy maybe?

**Comment by stapelberg** [comment]
I looked into this some more, and I’m still not sure how I should refactor the code to be able to put this into a separate plugin.

Also, consider that IPv4/IPv6 configuration, SSL and proxy support all live in the irc plugin. I consider RobustSession to operate on the same level, i.e. it’s “a thing that wraps actual IRC messages”.

Can anyone come up with some concrete suggestions with regards to how I can put the code into a separate plugin? If not, I’ll just put it into the irc plugin for now and send a pull request, so that we have some actual code to discuss.

**Comment by stapelberg** [comment]
Let’s move the discussion about how it should be implemented to https://github.com/weechat/weechat/pull/387
